service/tbc/utxoindex.go

```diff
@@ -0,0 +1,341 @@
+			}
+			op := tbcd.NewOutpoint(txIn.PreviousOutPoint.Hash,
+				txIn.PreviousOutPoint.Index)
+			if utxo, ok := utxos[op]; ok && !utxo.IsDelete() {
+				delete(utxos, op)
+				continue
+			}
+		}
+		for outIndex, txOut := range tx.MsgTx().TxOut {
+			if txscript.IsUnspendable(txOut.PkScript) {
+				continue
+			}
+			utxos[tbcd.NewOutpoint(*tx.Hash(), uint32(outIndex))] = tbcd.NewCacheOutput(
+				tbcd.NewScriptHashFromScript(txOut.PkScript),
+				uint64(txOut.Value),
+				uint32(outIndex))
+		}
+	}
+	return nil
+}
+
+func txOutFromOutPoint(ctx context.Context, db tbcd.Database, op tbcd.Outpoint) (*wire.TxOut, error) {
+	txId := op.TxIdHash()
+	txIndex := op.TxIndex()
+
+	// Find block hashes
+	blockHash, err := db.BlockHashByTxId(ctx, *txId)
+	if err != nil {
+		return nil, fmt.Errorf("block by txid: %w", err)
+	}
+	b, err := db.BlockByHash(ctx, *blockHash)
+	if err != nil {
+		return nil, fmt.Errorf("block by hash: %w", err)
+	}
+	for _, tx := range b.Transactions() {
+		if !tx.Hash().IsEqual(txId) {
+			continue
+		}
+		txOuts := tx.MsgTx().TxOut
+		if len(txOuts) < int(txIndex) {
+			return nil, fmt.Errorf("tx index invalid: %v", op)
+		}
+		return txOuts[txIndex], nil
+	}
+
+	return nil, fmt.Errorf("tx id not found: %v", op)
+}
+
+func unprocessUtxos(ctx context.Context, db tbcd.Database, block *btcutil.Block, utxos map[tbcd.Outpoint]tbcd.CacheOutput) error {
+	txs := block.Transactions()
+	// Walk backwards through the txs
+	for idx := len(txs) - 1; idx >= 0; idx-- {
+		tx := txs[idx]
+		// TxIn get data from disk and insert into the cache as insert
+		for _, txIn := range tx.MsgTx().TxIn {
+			if blockchain.IsCoinBase(tx) {
+				// Skip coinbase inputs
+				break
+			}
+
+			op := tbcd.NewOutpoint(txIn.PreviousOutPoint.Hash,
+				txIn.PreviousOutPoint.Index)
+			prevTxOut, err := txOutFromOutPoint(ctx, db, op)
+			if err != nil {
+				return fmt.Errorf("script value: %w", err)
+			}
+			// XXX this should not happen. We are keeping it for
+			// now to ensure it indeed does not happen. Remove in a
+			// couple of years.
+			if _, ok := utxos[op]; ok {
+				return fmt.Errorf("impossible collision: %v", op)
+			}
+			utxos[op] = tbcd.NewCacheOutput(tbcd.NewScriptHashFromScript(prevTxOut.PkScript),
+				uint64(prevTxOut.Value), txIn.PreviousOutPoint.Index)
+		}
+
+		// TxOut if those are in the cache delete from cache; if they
+		// are not in the cache insert "delete from disk command" into
+		// cache.
+		for outIndex, txOut := range tx.MsgTx().TxOut {
+			if txscript.IsUnspendable(txOut.PkScript) {
+				continue
+			}
+
+			op := tbcd.NewOutpoint(*tx.Hash(), uint32(outIndex))
+			if _, ok := utxos[op]; ok {
+				delete(utxos, op)
+			} else {
+				utxos[op] = tbcd.NewDeleteCacheOutput(tbcd.NewScriptHashFromScript(txOut.PkScript),
+					op.TxIndex())
+			}
+		}
+	}
+
+	return nil
+}
+
+func (s *Server) fetchOPParallel(ctx context.Context, c chan struct{}, w *sync.WaitGroup, op tbcd.Outpoint, utxos map[tbcd.Outpoint]tbcd.CacheOutput) {
```

Comment:
Is there a reason for all of the below functions to be receivers on the `Server` struct? I believe these should all be separated unless they are explicitly required by the Server as well as the UTXO indexer. If they are required by both, ideally they should not be receivers.

The only usage of `s` I am seeing is for `s.mtx`, which I don't believe is necessary in most cases.